main/src/ca/uwaterloo/flix/language/phase/UllmansAlgorithm.scala

```diff
@@ -158,39 +159,72 @@ object UllmansAlgorithm {
-      seen.add(curr)
-
-      // Recursively visit each unseen child.
-      for ((succ, loc) <- succ.getOrElse(curr, Set.empty)) {
-        if (!seen.contains(succ)) {
-          pred.update(succ, (curr, loc))
-          visit(succ)
+    // Find one of the shortest paths from `start` to `goal`
+    def bfs(src: Name.Pred, dst: Name.Pred): Option[mutable.Map[Name.Pred, (Name.Pred, SourceLocation)]] = {
+      // A map from predicates to one of their immediate
+      // predecessors in the BFS.
+      val pred = mutable.Map.empty[Name.Pred, (Name.Pred, SourceLocation)]
+      // A set of previously seen predicates.
+      val seen = mutable.Set.empty[Name.Pred]
+      // the tasklist for bfs
+      val tasklist = mutable.Queue.empty[Name.Pred]
+      seen.add(src)
+      tasklist.enqueue(src)
+
+      while (tasklist.nonEmpty) {
+        val current = tasklist.dequeue()
+        if (current == dst) return Some(pred)
+        succ.getOrElse(current, Set.empty).foreach {
+          case (next, ruleLoc) =>
+            if (!seen.contains(next)) {
+              pred.update(next, (current, ruleLoc))
+              seen.add(next)
+              tasklist.enqueue(next)
+            }
         }
       }
+      None
     }
 
-    // Compute the predecessor map.
-    visit(dst)
+    // Recursively constructs a path from `endPoint` and backwards through the graph.
+    def unroll(from: Name.Pred, to: Name.Pred, pred: mutable.Map[Name.Pred, (Name.Pred, SourceLocation)]): List[(Name.Pred, SourceLocation)] = {
+      @tailrec
+      def unrollHelper(s: Name.Pred, acc: List[(Name.Pred, SourceLocation)]): List[(Name.Pred, SourceLocation)] = {
+        if (s == to) return acc
+        pred.getOrElse(s, throw InternalCompilerException("Stratification cycle malformed")) match {
+          case (prev, loc) => unrollHelper(prev, (prev, loc) :: acc)
+        }
+      }
 
-    // Recursively constructs a path from `src` and backwards through the graph.
-    def unroll(curr: Name.Pred): List[(Name.Pred, SourceLocation)] = pred.get(curr) match {
-      case None => Nil
-      case Some((prev, loc)) => (prev, loc) :: unroll(prev)
+      unrollHelper(from, Nil).reverse
     }
 
-    // Assemble the full path.
-    (src, loc) :: unroll(src) ::: (src, loc) :: Nil
+    // We do not know where the cycle is but often it includes the edge
+    // `firstCheck` since that edge brought ullman over the limit so
+    // we try to find that cycle first and then check every other strong edge.
+    bfs(firstCheck.head, firstCheck.body) match {
+      case Some(pred) =>
+        // we found a cycle and can report it
+        val DependencyEdge.Strong(head, body, loc) = firstCheck
+        (body, loc) :: unroll(body, head, pred) ::: (body, loc) :: Nil
+      case None =>
+        // `firstCheck` is not a part of a cycle to we check the same
+        // for all strong edges.
+        // TODO: The strong cycle can reach `firstcheck.head` so we could
```

Comment:
We tend to avoid TODOs in code, especially for less important things. Instead you should open a ticket! 

TODOs in code are forgotten. Less so for tickets.